uvicorn/protocols/websockets/wsproto_impl.py

```diff
@@ -125,9 +125,12 @@ def resume_writing(self):
         self.writable.set()
 
     def shutdown(self):
-        self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
-        output = self.conn.send(wsproto.events.CloseConnection(code=1012))
-        self.transport.write(output)
+        if self.handshake_complete:
+            self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
+            output = self.conn.send(wsproto.events.CloseConnection(code=1012))
+            self.transport.write(output)
```

Comment:
This is the core here.

We want to check if the handshake was completed, because `CloseConnection` is not a valid event for `wsproto.WSConnection.send()` when the handshake is not completed.

This is the error these lines solve:
```bash
wsproto.utilities.LocalProtocolError: Event CloseConnection(code=1012, reason=None) cannot be sent during the handshake
```

More about it on #596.